### PR TITLE
feat(auth): Tenant token

### LINKED ISSUE #1991: Tenant Tokens (Multi-Tenant indexes)
This issue is tightly linked to this WIP spec: https://github.com/meilisearch/specifications/pull/89

Check the spec and the PR description (#2096) for more details

### PR DESCRIPTION
**Pull request by ManyTheFish**
Make meilisearch support JWT authentication signed with meilisearch API keys
using HS256, HS384 or HS512 algorithms.

Related spec: [specifications#89](https://github.com/meilisearch/specifications/pull/89) [rendered](https://github.com/meilisearch/specifications/blob/scoped-api-keys/text/0089-tenant-tokens.md)
Fix #1991

### CODE REVIEW

File: meilisearch-http/src/routes/api_key.rs [hunk 1]
```diff
@@ -29,8 +29,8 @@ pub async fn create_api_key(
     body: web::Json<Value>,
     _req: HttpRequest,
 ) -> Result<HttpResponse, ResponseError> {
-    let key = auth_controller.create_key(body.into_inner()).await?;
```
Comment (MarinPostma) [inline meilisearch-http/src/routes/api_key.rs:32]: No!! this function in blocking, you can't call it in this context! You need to make them async again, and call the store operation from a `tokio::spawn_blocking` context

File: meilisearch-http/src/extractors/authentication/mod.rs [hunk 2]
```diff
@@ -126,15 +141,82 @@ pub mod policies {
                 return Some(AuthFilter::default());
             }
 
-            // authenticate if token is allowed.
-            if let Some(action) = Action::from_repr(A) {
+            if token.contains('.') {
+                // Tenant token
+                return ActionPolicy::<A>::authenticate_tenant_token(auth, token, index);
```
Comment (MarinPostma) [inline meilisearch-http/src/extractors/authentication/mod.rs:146]: why is that sufficient?

**Review by MarinPostma — CHANGES_REQUESTED** [review]

File: meilisearch-http/src/routes/api_key.rs [hunk 3]
```diff
@@ -66,9 +66,8 @@ pub async fn patch_api_key(
 ) -> Result<HttpResponse, ResponseError> {
     let key = auth_controller
         // keep 8 first characters that are the ID of the API key.
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/api_key.rs]: What does this comment refer to?

File: meilisearch-http/src/extractors/authentication/mod.rs (hunk 2 above)
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:146]: API keys can't contain `.`, the master key could but it's checked before. We can change this into:
```suggestion
            if let Some(filters) = ActionPolicy::<A>::authenticate_tenant_token(auth, token, index) {
                // Tenant token
                return Some(filters);
```

which will try to authenticate to know if it is a tenant token or not. 🙂

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Here is half of my review as you updated the branche.

File: meilisearch-auth/src/lib.rs [hunk 4]
```diff
@@ -97,47 +119,136 @@ impl AuthController {
...
     }
+
+    /// Check if the provided key is valid
+    /// and is authorized to make a specific action.
+    pub fn authenticate(&self, key: &[u8], action: Action, index: Option<&[u8]>) -> Result<bool> {
+        if self.is_key_authorized(key, action, index)? {
+            self.is_key_valid(key)
+        } else {
+            Ok(false)
+        }
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-auth/src/lib.rs:176]: in `is_key_authorized` we have LMDB accesses to check key authorization, in `is_key_valid` we have a sha256 digest, I chose to do the hash at the end. I can change the order if you want. ☺️

File: meilisearch-http/src/extractors/authentication/mod.rs [hunk 5]
```diff
@@ -126,15 +141,82 @@ pub mod policies {
...
     }
+
+    impl<const A: u8> ActionPolicy<A> {
+        fn authenticate_tenant_token(
+            auth: AuthController,
+            token: &str,
+            index: Option<&str>,
+        ) -> Option<AuthFilter> {
+            // Only search action can be accessed by a tenant token.
+            if A != actions::SEARCH {
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:166]: A is the action needed to be authorized to access a specific route, for example:
https://github.com/meilisearch/MeiliSearch/blob/main/meilisearch-http/src/routes/indexes/documents.rs#L159

File: meilisearch-http/src/extractors/authentication/mod.rs [hunk 6]
```diff
@@ -50,19 +50,23 @@ impl<P: Policy + 'static, D: 'static + Clone> FromRequest for GuardedData<P, D>
                     Some("Bearer") => {
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: Only "Bearer" written with a capital is accepted, with exactly one space between it and the following key.

[the remark this answers is not in the document]
Comment (MarinPostma) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: uhm that is not http compliant: https://stackoverflow.com/a/5259004

File: meilisearch-auth/src/lib.rs [hunk 7]
```diff
@@ -97,47 +119,136 @@ impl AuthController {
...
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Default, Clone)]
+pub struct IndexSearchRules {
+    pub filter: Option<serde_json::Value>,
+}
+
+fn generate_key(master_key: &[u8], uid: &str) -> String {
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-auth/src/lib.rs:250]: Ouch, no, you're right, it's bad naming, it should be keyid

File: meilisearch-http/src/extractors/authentication/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-http/src/extractors/authentication/mod.rs]: That's cool 😮

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
bors try

File: meilisearch-http/src/extractors/authentication/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: @MarinPostma @Kerollmops, "Bearer" is not the Header name but a part of the Value, the header name is "Authentication".

[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: I don't know if it's the good RFC, but, here is a pseudo regex matching the bearer auth:
https://datatracker.ietf.org/doc/html/rfc6750#section-2.1

File: meilisearch-lib/src/index/search.rs [hunk 8]
```diff
@@ -100,11 +106,32 @@ impl Index {
...
+                        Box::new(FilterCondition::from(filter)),
+                    );
+                    Some(Filter::from(fc))
+                }
+                None => Some(f),
+            };
+        }
+
+        if let Some(filter) = filter {
+            search.filter(filter);
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-lib/src/index/search.rs]: Done @MarinPostma, I could create a function in an `impl SearchQuery`, but I chose to implement a standalone function in order to keep everything in meilisearch-http:
https://github.com/meilisearch/meilisearch/blob/7ca647f0d0adf18cb7a05a91ab60a18cc2ff9beb/meilisearch-http/src/routes/indexes/search.rs#L83-L101

**Comment by curquiza** [comment]
bors try

File: meilisearch-lib/src/index_controller/mod.rs [hunk 9]
```diff
@@ -543,17 +544,14 @@ where
         Ok(stats)
     }
 
-    pub async fn get_all_stats(&self, index_filter: &Option<Vec<String>>) -> Result<Stats> {
+    pub async fn get_all_stats(&self, search_rules: &SearchRules) -> Result<Stats> {
```
Comment (MarinPostma) [inline meilisearch-lib/src/index_controller/mod.rs:547]: I'll find a way to not depend on meilisearch auth here later, so that's ok for now

**Review by MarinPostma — APPROVED** [review]

File: meilisearch-http/src/extractors/authentication/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: If it is a real issue can you "Reference in a new issue", please? If it is not, let's merge that and pray!

**Comment by Kerollmops** [comment]
bors merge

File: meilisearch-http/src/extractors/authentication/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:50]: There is no issue here, I don't understand your point.

**Review by Kerollmops — APPROVED** [review]
bors merge

**Pull request merged**